[Jan Alfenson Tan] iP

Cyolune opened the pull request:
# Sana is here!
> "Sana is BEEEEG." - Sana ([source](https://www.youtube.com/channel/UCsUj0dszADCGbF3gNrQEuSQ))

Sana is here to help you keep track of things. She's
- ~text~-based
- pretty fast considering her size
- kinda cute ngl

All you need to do is,
1. download it from [here](https://github.com/Cyolune/ip/releases/tag/A-Jar)
2. run it on the command line
3. play!

And its **FREE**! :star_struck:

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too! Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
````

Commit:
Add Gradle support

Cyolune committed:
Rename Duke to Sana
Add Greet, Echo, Exit functionality

Cyolune committed:
Rename Duke to Sana, Add Greet, Echo, Bye functionality

Cyolune committed:
Define Sana object
Add greet method and static border

Cyolune committed:
Add methods to Sana, Add store user input functionality

Cyolune committed:
Add Task class

Cyolune committed:
Change Sana to store Tasks

Cyolune committed:
Add mark and unmark support to Sana

Cyolune committed:
Change Sana to use toString method of task
Add printing of task when marking/unmarking

Cyolune committed:
Change getTaskName to toString

Cyolune committed:
Add ToDo class

Cyolune committed:
Add ToDo support to Sana

Cyolune committed:
Change Task toString to return task name only

Cyolune committed:
Add Event class

Cyolune committed:
Add support for Events + Change list number text

Cyolune committed:
Add Deadline class + Rename variables in Event

Cyolune committed:
Add support for Deadlines in Sana

Cyolune committed:
Change text-ui-test input and EXPECTED

Cyolune committed:
Add UnknownCommandException, Add support for exception in Sana

Cyolune committed:
Add IncompleteCommandException

Cyolune committed:
Add support for IncompleteCommandException in Sana

Cyolune committed:
Add OutOfBoundsMark exception class

Cyolune committed:
Add OutOfBoundMarkException + Add support in Sana. Add documentation to IncompleteCommandException

Cyolune committed:
Add Delete support to Sana, Rename OutOfBoundsMarkException to OutOfBoundsTaskException

Cyolune committed:
Add feature1.txt

Cyolune committed:
Add feature2

Cyolune committed:
Merge branch 'master' into feature1

Cyolune committed:
Readd feature1.txt

Cyolune committed:
Edit feature2.txt

Cyolune committed:
Merge branch 'master' into feature1

Cyolune committed:
Remove feature1 and feature2

Cyolune committed:
Remove feature1.txt

Cyolune committed:
Add Memory.java class + Add support for Memory in Sana

Cyolune committed:
Add mem reading functionality

Cyolune committed:
Add support to memory writing in Sana

Cyolune committed:
Replace mem splitter from | to :

Cyolune committed:
Remove redundant getName method from task

Cyolune committed:
Add clarifying comment in memory class

Cyolune committed:
Add support for LocalDate in Sana

Cyolune committed:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/Sana.java

Cyolune committed:
Update sana.txt to store dates in proper format

Cyolune committed:
Add ui class

Cyolune committed:
Move most ui commands from sana to ui

Cyolune committed:
Shift more print methods to ui for better abstraction

Cyolune committed:
Readjust import statement positions

Cyolune committed:
Add tasklist class and support for tasklist class in sana

Cyolune committed:
Add Parser class

Cyolune committed:
Add support for Parser in Sana

Cyolune committed:
Organise files to packages

Cyolune committed:
Add JUnit Tests

Cyolune committed:
Add JAR release

Cyolune committed:
Add javadocs to all classes

Cyolune committed:
Fix code to follow Java Coding Standard

Cyolune committed:
Add Find support

Cyolune committed:
Merge branch 'branch-A-CodingStandard'

Cyolune committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/sana/Sana.java
#	src/main/java/sana/exception/IncompleteCommandException.java
#	src/main/java/sana/exception/OutOfBoundsTaskException.java
#	src/main/java/sana/exception/UnknownCommandException.java

Cyolune committed:
Fix minor documentation mistakes

Cyolune committed:
Git stuff

Cyolune committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Cyolune committed:
Add checkstyle test and fix checkstyle issues

Cyolune committed:
Add GUI functionality to Sana

TypeDefinition commented on src/main/java/sana/Memory.java:
```diff
@@ -0,0 +1,103 @@
+                e.printStackTrace();
+            }
+        }
+        this.memFile = new File(MEM_PATH);
```
I noticed that the `this` keyword is not used when referring to `memFile` anywhere else in this function. Perhaps consider making this line consistent too?

TypeDefinition commented on src/main/java/sana/task/Task.java:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @param done  sets the sana.task's completion
+     */
+    public void setDone(boolean done) {
+        this.isDone = done;
+    }
```
As per the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#naming), perhaps consider renaming `done` to `isDone`?

TypeDefinition reviewed (commented):
You've followed the coding standard very well! I struggled to find any mistakes at all! I also find the incorporation of Space Australian English into the bot's replies very cute. :australia: :smiley:

Cyolune committed:
Encapsulate commands to their own classes

Cyolune committed:
Remove redundant methods

Cyolune committed:
Remove system print lines in parser

CS2103Lerulatus commented on src/main/java/sana/Sana.java:
```diff
@@ -58,29 +50,30 @@ public String doCommand(String userCommand) {
-                    return new FindCommand().executeCommand(parsedCmd, userTasks);
-                default:
-                    throw new UnknownCommandException();
+            case "bye":
+                return new ByeCommand().executeCommand(parsedCmd, userTasks);
+            case "list":
+                return new ListCommand().executeCommand(parsedCmd, userTasks);
+            case "mark":
+                return new MarkCommand().executeCommand(parsedCmd, userTasks);
+            case "unmark":
+                return new UnmarkCommand().executeCommand(parsedCmd, userTasks);
+            case "todo":
+            case "event":
+            case "deadline":
+                return new AddTask().executeCommand(parsedCmd, userTasks);
+            case "delete":
+                return new DeleteCommand().executeCommand(parsedCmd, userTasks);
+            case "find":
+                return new FindCommand().executeCommand(parsedCmd, userTasks);
+            default:
+                throw new UnknownCommandException();
```
——— For PR review ———
May I suggest the further use of OOP-styles -- package the case statements into the Command object itself. A `ByeCommand` object should contain the alias `"bye"` that trigger it's execution, for example.

Most commands would benefit from having multiple aliases that trigger their execution too, such as `bye, exit, quit` all being valid and recognized commands that executes the `ByeCommand` action. Writing all these aliases into a switch-case would prove troublesome, especially as the project scales and expands to include new commands and features, so packing them into their respective Command object would prove to be handy for future scaling.

Also, this switch-case block follows the recommended indentation, so that's good.

CS2103Lerulatus commented on src/main/java/sana/Memory.java:
```diff
@@ -0,0 +1,103 @@
+package sana;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.Scanner;
```
——— For PR review ———
Nice and clear use of imports, even when `import java.io.*` is tempting to write in projects that extensively use io functionality.

Cyolune committed:
Add gui improvements

Cyolune committed:
Add assertions to code

High density of assert statements in code can help to reduce potential
in code.

Let's add some assert statements to our code to ensure minimal bugs as
the project grows in size.

Cyolune committed:
Remove sanaResponse import in MarkCommand

Cyolune committed:
Refactor code to improve code quality and readability

There are many benefits to improving code quality, which includes
easier maintenance of the codebase in the future in the event others
want to make changes to the program.

Some boolean checks were refactored to be self-explanatory. Comments
were added to some blocks of codes and methods to add clarity to the
code.

Some improvements in method length in the Parser class could be
good, but were not implemented in this commit yet as research is still
being done on the best way to refactor it.

Cyolune committed:
Merge pull request #2 from Cyolune/branch-A-CodeQuality

Refactor code to improve code quality and readability

Cyolune committed:
Merge pull request #3 from Cyolune/branch-A-Assertions

Branch a assertions

Add assertions to code

High density of assert statements in code can help to reduce potential
in code.

Let's add some assert statements to our code to ensure minimal bugs as
the project grows in size.

Cyolune committed:
Add update functionality to sana

Refactor exceptions to extend SanaException, a parent class for all
exceptions Sana throws

Cyolune committed:
Update exceptions to use SanaExceptions

Cyolune committed:
Improve GUI, readd Level-10 tag to do merge-level-10 task

Cyolune committed:
Fix Update date parsing bug

Cyolune committed:
Add close window functionality to bye

Cyolune committed:
Add Ui.png to docs

Cyolune committed:
Update README.md

Cyolune committed:
Update README.md

Cyolune committed:
Add memory corruption fix

Cyolune committed:
Update README.md
